### PR TITLE
[flang] Improve disjoint/identical slices recognition in opt-bufferization.

### PR DESCRIPTION
**Pull request by vzakhari**
The changes are needed to be able to optimize 'x(9,:)=SUM(x(1:8,:),DIM=1)'
without a temporary array. This pattern exists in exchange2.

The patch also fixes an existing problem in Flang with this test:
```
program main
  integer :: a(10) = (/1,2,3,4,5,6,7,8,9,10/)
  integer :: expected(10) = (/1,10,9,8,7,6,5,4,3,2/)
  print *, 'INPUT: ', a
  print *, 'EXPECTED: ', expected
  call test(a, 10, 2, 10, 9)
  print *, 'RESULT: ', a
contains
  subroutine test(a, size, x, y, z)
    integer :: x, y, z, size
    integer :: a(:)
    a(x:y:1) = a(z:x-1:-1) + 1
  end subroutine test
end program main
```

### CODE REVIEW

**Review by tblah — APPROVED** [review]
This is _really_ nice

File: flang/lib/Optimizer/HLFIR/Transforms/OptimizedBufferization.cpp [hunk 1]
```diff
@@ -159,28 +159,162 @@ containsReadOrWriteEffectOn(const mlir::MemoryEffects::EffectInstance &effect,
...
+  // If stride is known to be non-negative, then the ordered
+  // bounds match the <lb, ub> of the descriptor.
+  // If stride is known to be negative, then the ordered
+  // bounds are <ub, lb> of the descriptor.
+  // If stride is unknown, we cannot deduce any order,
+  // so the result is <nullptr, nullptr>
+  static std::pair<mlir::Value, mlir::Value>
+  getOrderedBounds(const SectionDesc &desc) {
+    mlir::Value stride = desc.stride;
+    // Null stride means stride-1.
```
Comment (jeanPerier) [inline flang/lib/Optimizer/HLFIR/Transforms/OptimizedBufferization.cpp:244]:
```suggestion
    // Null stride means stride=1.
```

**Review by jeanPerier — APPROVED** [review]
Looks great, thanks!

I wonder if the isLess could be improved/made even more generic by using MLIR/LLVM Presburger (I have never used it, so this is a random comment, and I do not expect a follow-up here).

**Comment by vzakhari** [comment]
> Looks great, thanks!
> 
> I wonder if the isLess could be improved/made even more generic by using MLIR/LLVM Presburger (I have never used it, so this is a random comment, and I do not expect a follow-up here).

Thanks for the hint, Jean! I will learn about MLIR Presburger to see how it can be applied.

**Pull request merged**